packages/mui-base/src/utils/useRenderPropForkRef.ts

```diff
@@ -8,7 +10,13 @@ import { useForkRef } from './useForkRef';
 export function useRenderPropForkRef<ElementType extends React.ElementType, OwnerState>(
   render: BaseUIComponentProps<ElementType, OwnerState>['render'],
   ...refs: Array<React.Ref<any>>
-) {
-  const childRef = typeof render !== 'function' ? render.ref : null;
+): React.RefCallback<any> | null {
+  let childRef;
+  if (isReactVersionAtLeast(19)) {
+    childRef = typeof render !== 'function' ? render.props.ref : null;
+  } else {
+    childRef = typeof render !== 'function' ? render.ref : null;
+  }
```

Comment:
Better fork after the `typeof`. But also not sure this is wise: https://github.com/mui/material-ui/pull/42818#discussion_r1664248303. The explore we are exploring in https://github.com/mui/material-ui/pull/43022. To decide then use the same solution through the codebase.

Reply:
> Better fork after the typeof

You mean `const childRef = typeof render !== 'function' ? isReactVersionAtLeast(19) ? render.props.ref : render.ref : null;`? This makes it a nested ternary which I believe we disallow in our ESLint rules.

> I'm sure this is best: https://github.com/mui/material-ui/pull/42818#discussion_r1664248303.

How do you measure "best"? The version check is faster than property check (https://jsbench.me/mrm16egz7n/1) and the function call explains pretty well that we're branching because something changed in React version.

Reply:
I think that we should reverse  the condition, from

```js
  let childRef;
  if (isReactVersionAtLeast(19)) {
    childRef = typeof render !== 'function' ? render.props.ref : null;
  } else {
    childRef = typeof render !== 'function' ? render.ref : null;
  }
```

to 

```js
  let childRef;

  if (typeof render !== 'function') {
    childRef = isReactVersionAtLeast(19) ? render.props.ref : render.ref;
  } else {
    childRef = null;
  }
```

Reply:
> I was proposing to go...

All right, makes sense :+1:

> The only ask is that we are moving to a place where Base UI is exporting one helper and Material UI, etc. using it. I don't think that this should be duplicated.

I see your point. I don't feel comfortable sharing Base UI utils just yet, though, even internally. When we're closer to the release and breaking changes are less likely, we can review the whole codebase and move utilities where they ultimately belong.

Reply:
> I don't feel comfortable sharing Base UI utils just yet, though, even internally

We could duplicate the logic until this happen. The outcome I'm trying to go after is: learn once fix everywhere.